Upgrade webrtc-audio-processing lib to v1.0

skywhale opened the pull request:
Making this as a draft until I do a bit more testing, but let me share it for an early peak and feedback.

strohel commented on webrtc-audio-processing-sys/build.rs:
```diff
@@ -16,8 +11,9 @@ mod webrtc {
     use failure::bail;
 
     const LIB_NAME: &str = "webrtc-audio-processing";
+    const LIB_MIN_VERSION: &str = "1.0.0";
```
```suggestion
    const LIB_NAME: &str = "webrtc-audio-processing-1";
    const LIB_MIN_VERSION: &str = "1.0";
```

I had to make these changes for this to find my installed webrtc-audio-processing 1.0. But I made the distro package myself, and no other distro seems to package webrtc-audio-processing 1.0 yet, so I cannot confirm how the pkgconfig file name and version will look like.

skywhale replied:
I checked the latest pkgconfig and they are actually naming it "webrtc-audio-processing-1". Thanks for the catch :)

strohel commented on webrtc-audio-processing-sys/src/lib.rs:
```diff
@@ -1,9 +1,13 @@
 #![allow(non_upper_case_globals)]
 #![allow(non_camel_case_types)]
 #![allow(non_snake_case)]
+// https://github.com/rust-lang/rust-bindgen/issues/1651
+#![allow(deref_nullptr)]
```
Thanks for this - we should add it to our other -sys packages, too.

strohel reviewed (commented):
Whoa, this must have been a ton of work! I've reviewed roughly the first half of the expanded diff, added some comments. I haven't worked with the library yet, so please take them with a grain of salt.

Thanks for taking care to update and extend tests, that's admirable you found energy to do that despite overall size of the change.

bschwind commented on src/config.rs:
```diff
@@ -1,135 +1,133 @@
-    /// Higher suppression level.
-    High,
+pub struct Pipeline {
+    /// Maximum allowed processing rate used internally. May only be set to 32000 or 48000 and any
+    /// differing values will be treated as 48000. The default rate is currently selected based on
+    /// the CPU architecture.
+    pub maximum_internal_processing_rate: u32,
```
Agreed here, this should be restricted if those are the only two values that make sense.

skywhale replied:
Done.

skywhale replied:
Done.

skywhale commented:
I gave up on [this TODO](https://github.com/tonarino/webrtc-audio-processing/blob/330f6fe09ad1f23f91b0db1b22f5fa87342cb5a7/webrtc-audio-processing-sys/src/wrapper.hpp#L5-L6). Bindgen cannot bind virtual functions (https://github.com/rust-lang/rust-bindgen/issues/27).

skywhale commented on src/config.rs:
```diff
@@ -1,135 +1,133 @@
+/// A configuration for initializing a Processor instance.
+#[derive(Debug, Clone, PartialEq)]
+pub struct InitializationConfig {
+    /// Number of the input and output channels for the capture frame.
+    pub num_capture_channels: usize,
+    /// Number of the input and output channels for the render frame.
+    pub num_render_channels: usize,
```
`capture > render` is the order I see in the original API. Fixed.

skywhale commented on src/config.rs:
```diff
@@ -1,135 +1,133 @@
-    /// ignored when this option is set.
-    pub stream_delay_ms: Option<i32>,
+pub struct PreAmplifier {
+    /// Fixed linear gain multiplifier. The default has no effect.
```
Done.

skywhale commented on src/config.rs:
```diff
@@ -1,135 +1,133 @@
+    /// Export the output of linear AEC for custom processing.
+    pub export_linear_aec_output: bool,
+    /// Enforce the highpass filter to be on. It has no effect for the mobile mode.
+    pub enforce_high_pass_filtering: bool,
```
Done.

skywhale commented on src/config.rs:
```diff
@@ -158,118 +160,221 @@ impl From<NoiseSuppressionLevel> for ffi::NoiseSuppression_SuppressionLevel {
+    /// Sets the maximum gain the digital compression stage may apply, in dB. A higher number
+    /// corresponds to greater compression, while a value of 0 will leave the signal uncompressed.
+    /// Limited to [0, 90]. For updates after APM setup, use a RuntimeSetting instead.
+    pub compression_gain_db: u32,
```
Done.

skywhale commented on src/config.rs:
```diff
@@ -158,118 +160,221 @@ impl From<NoiseSuppressionLevel> for ffi::NoiseSuppression_SuppressionLevel {
+    pub enable_residual_echo_detector: bool,
+
+    /// Enables reporting of [`output_rms_dbfs`] in [`Stats`].
+    pub enable_level_estimation: bool,
```
A good idea :) Done.

skywhale commented on src/config.rs:
```diff
@@ -158,118 +160,221 @@ impl From<NoiseSuppressionLevel> for ffi::NoiseSuppression_SuppressionLevel {
-            enabled_value.into()
+impl From<Config> for ffi::AudioProcessing_Config {
+    fn from(other: Config) -> Self {
+        let pre_amplifier = if let Some(config) = other.pre_amplifier {
+            config.into()
         } else {
```
Yeah, I sort of like that each field has the explicit `enabled: false`, esp because it's a FFI struct that doesn't follow Rust conventions. I think I will leave it as is?

skywhale commented on src/config.rs:
```diff
@@ -278,64 +383,55 @@ impl From<Config> for ffi::Config {
+    /// capture frame, after processing. It is constrained to [-127, 0]. The computation follows:
+    /// https://tools.ietf.org/html/rfc6465 with the intent that it can provide the RTP audio level
+    /// indication. Only reported if level estimation is enabled in [`Config`].
+    pub output_rms_dbfs: Option<i32>,
```
Done.

skywhale commented on src/config.rs:
```diff
@@ -278,64 +383,55 @@ impl From<Config> for ffi::Config {
-    /// every second.
+    /// The delay standard deviation in milliseconds. The values are aggregated until the first
+    /// call to [`get_stats()`] and afterwards aggregated and updated every second.
     pub delay_standard_deviation_ms: Option<i32>,
```
Done.

bschwind commented:
@skywhale are you able to build this on MacOS? I'm getting the following error:

```
<bunch of CMake output>
...
-- Installing: /Users/brian/projects/tonari/webrtc-audio-processing/target/release/build/webrtc-audio-processing-sys-0357fd24f02783fb/out/lib/libabsl_time_zone.a
-- Installing: /Users/brian/projects/tonari/webrtc-audio-processing/target/release/build/webrtc-audio-processing-sys-0357fd24f02783fb/out/lib/libabsl_bad_any_cast_impl.a
-- Installing: /Users/brian/projects/tonari/webrtc-audio-processing/target/release/build/webrtc-audio-processing-sys-0357fd24f02783fb/out/lib/libabsl_bad_optional_access.a
-- Installing: /Users/brian/projects/tonari/webrtc-audio-processing/target/release/build/webrtc-audio-processing-sys-0357fd24f02783fb/out/lib/libabsl_bad_variant_access.a
cargo:root=/Users/brian/projects/tonari/webrtc-audio-processing/target/release/build/webrtc-audio-processing-sys-0357fd24f02783fb/out/./abseil-cpp

--- stderr
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_ASM_COMPILER
    CMAKE_ASM_FLAGS
    CMAKE_C_COMPILER
    CMAKE_C_FLAGS


/Users/brian/projects/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/time/internal/cctz/src/time_zone_info.cc:721:35: warning: declaration shadows a local variable [-Wshadow-uncaptured-local]
      name, [](const std::string& name) -> std::unique_ptr<ZoneInfoSource> {
                                  ^
/Users/brian/projects/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/time/internal/cctz/src/time_zone_info.cc:709:44: note: previous declaration is here
bool TimeZoneInfo::Load(const std::string& name) {
                                           ^
/Users/brian/projects/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/time/internal/cctz/src/time_zone_info.cc:722:18: warning: declaration shadows a local variable [-Wshadow-uncaptured-local]
        if (auto zip = FileZoneInfoSource::Open(name)) return zip;
                 ^
/Users/brian/projects/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/time/internal/cctz/src/time_zone_info.cc:720:8: note: previous declaration is here
  auto zip = cctz_extension::zone_info_source_factory(
       ^
/Users/brian/projects/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/time/internal/cctz/src/time_zone_info.cc:723:18: warning: declaration shadows a local variable [-Wshadow-uncaptured-local]
        if (auto zip = AndroidZoneInfoSource::Open(name)) return zip;
                 ^
/Users/brian/projects/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/time/internal/cctz/src/time_zone_info.cc:720:8: note: previous declaration is here
  auto zip = cctz_extension::zone_info_source_factory(
       ^
/Users/brian/projects/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/time/internal/cctz/src/time_zone_libc.cc:166:22: warning: declaration shadows a local variable [-Wshadow]
        if (std::tm* tmp = local_time(&lo, &tm)) {
                     ^
/Users/brian/projects/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/time/internal/cctz/src/time_zone_libc.cc:156:18: note: previous declaration is here
    if (std::tm* tmp = local_time(&mid, &tm)) {
                 ^
/Users/brian/projects/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/time/internal/cctz/src/time_zone_libc.cc:191:9: warning: result of comparison 'const std::int_fast64_t' (aka 'const long long') < -9223372036854775808 is always false [-Wtautological-type-limit-compare]
  if (s < std::numeric_limits<std::time_t>::min()) {
      ~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/brian/projects/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/time/internal/cctz/src/time_zone_libc.cc:195:9: warning: result of comparison 'const std::int_fast64_t' (aka 'const long long') > 9223372036854775807 is always false [-Wtautological-type-limit-compare]
  if (s > std::numeric_limits<std::time_t>::max()) {
      ~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/brian/projects/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/time/internal/cctz/src/time_zone_libc.cc:84:6: warning: unused function template 'tm_gmtoff' [-Wunused-template]
auto tm_gmtoff(const T& tm) -> decltype(tm.__tm_gmtoff) {
     ^
/Users/brian/projects/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/time/internal/cctz/src/time_zone_libc.cc:100:6: warning: unused function template 'tm_zone' [-Wunused-template]
auto tm_zone(const T& tm) -> decltype(tm.__tm_zone) {
     ^
5 warnings generated.
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: libabsl_debugging_internal.a(elf_mem_image.cc.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: libabsl_debugging_internal.a(vdso_support.cc.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: libabsl_debugging_internal.a(elf_mem_image.cc.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: libabsl_debugging_internal.a(vdso_support.cc.o) has no symbols
3 warnings generated.
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: libabsl_flags.a(flag.cc.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: libabsl_flags.a(flag.cc.o) has no symbols
warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: warning for library: libabsl_flags.a the table of contents is empty (no object file members in the library define global symbols)
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/brian/projects/tonari/webrtc-audio-processing/target/release/build/webrtc-audio-processing-sys-0357fd24f02783fb/out/lib/libabsl_debugging_internal.a(elf_mem_image.cc.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/brian/projects/tonari/webrtc-audio-processing/target/release/build/webrtc-audio-processing-sys-0357fd24f02783fb/out/lib/libabsl_debugging_internal.a(vdso_support.cc.o) has no symbols
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/brian/projects/tonari/webrtc-audio-processing/target/release/build/webrtc-audio-processing-sys-0357fd24f02783fb/out/lib/libabsl_flags.a(flag.cc.o) has no symbols
warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: warning for library: /Users/brian/projects/tonari/webrtc-audio-processing/target/release/build/webrtc-audio-processing-sys-0357fd24f02783fb/out/lib/libabsl_flags.a the table of contents is empty (no object file members in the library define global symbols)
Error: Os { code: 2, kind: NotFound, message: "No such file or directory" }
```

bschwind commented:
Oh, this is caused by me not having `meson` installed. I swear Rust's file handling error message is the cause of way too much misery:  `Error: Os { code: 2, kind: NotFound, message: "No such file or directory" }`

skywhale commented:
Ah yes, meson and ninja are two build tools you need. I've only tested on my macbook so far; there may be some more work necessary on Linux.

bschwind commented:
I'm getting more build errors on linux:

```
  --- stderr
  /home/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/synchronization/internal/graphcycles.cc: In member function ‘void absl::lts_2020_02_25::synchronization_internal::GraphCycles::RemoveNode(void*)’:
  /home/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/synchronization/internal/graphcycles.cc:451:26: error: ‘numeric_limits’ is not a member of ‘std’
    451 |   if (x->version == std::numeric_limits<uint32_t>::max()) {
        |                          ^~~~~~~~~~~~~~
  /home/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/synchronization/internal/graphcycles.cc:451:49: error: expected primary-expression before ‘>’ token
    451 |   if (x->version == std::numeric_limits<uint32_t>::max()) {
        |                                                 ^
  /home/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/synchronization/internal/graphcycles.cc:451:52: error: ‘::max’ has not been declared; did you mean ‘std::max’?
    451 |   if (x->version == std::numeric_limits<uint32_t>::max()) {
        |                                                    ^~~
        |                                                    std::max
  In file included from /usr/include/c++/11.1.0/algorithm:62,
                   from /home/tonari/webrtc-audio-processing/webrtc-audio-processing-sys/abseil-cpp/absl/synchronization/internal/graphcycles.cc:38:
  /usr/include/c++/11.1.0/bits/stl_algo.h:3467:5: note: ‘std::max’ declared here
   3467 |     max(initializer_list<_Tp> __l, _Compare __comp)
        |     ^~~
  make[2]: *** [absl/synchronization/CMakeFiles/graphcycles_internal.dir/build.make:76: absl/synchronization/CMakeFiles/graphcycles_internal.dir/internal/graphcycles.cc.o] Error 1
  make[1]: *** [CMakeFiles/Makefile2:2150: absl/synchronization/CMakeFiles/graphcycles_internal.dir/all] Error 2
  make[1]: *** Waiting for unfinished jobs....
  make: *** [Makefile:136: all] Error 2
  thread 'main' panicked at '
  command did not execute successfully, got: exit status: 2

  build script failed, must exit now', /home/tonari/.cargo/registry/src/github.com-1ecc6299db9ec823/cmake-0.1.45/src/lib.rs:894:5
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Possibly related? https://github.com/abseil/abseil-cpp/issues/832

skywhale commented:
This PR has been rebased on `webrtc-audio-processing` v1.3.

morajabi commented:
@skywhale I'm getting an error message trying to use this branch with bundled flag on macOS: 

```   cargo:warning=In file included from src/wrapper.cpp:1:

  cargo:warning=In file included from src/wrapper.hpp:8:

  cargo:warning=/Users/mo/dev/noor/target/debug/build/webrtc-audio-processing-sys-4ab82baded4bbe6f/out/include/webrtc-audio-processing-1/modules/audio_processing/include/audio_processing.h:26:10: fatal error: 'absl/types/optional.h' file not found

  cargo:warning=#include "absl/types/optional.h"

  cargo:warning=         ^~~~~~~~~~~~~~~~~~~~~~~

  cargo:warning=1 error generated.

  exit status: 1

  --- stderr```

jacksongoode commented:
Just curious why the change was made to remove the abseil-cpp builder? I've found, that at least on my machine (macOS 15) with all the build requirements, that I get:

```
The system library `webrtc-audio-processing-1` required by crate `webrtc-audio-processing-sys` was not found.
The file `webrtc-audio-processing-1.pc` needs to be installed and the PKG_CONFIG_PATH environment variable must contain its parent directory.
The PKG_CONFIG_PATH environment variable is not set.

HINT: if you have installed the library, try setting PKG_CONFIG_PATH to the directory containing `webrtc-audio-processing-1.pc`.
```

I've made a branch to link the abseil library back and it built successfully. I can open a PR to merge into this branch.

skywhale commented:
@jacksongoode Thank you for investigating this and opening up a PR :)

> Just curious why the change was made to remove the abseil-cpp builder?

I removed it because the new build system using `meson` pulls the dependency automatically and installs it under the build directory, which we can reference. I believe in your and @morajabi's case, `meson` found a system-installed abseil-cpp and skipped the step to pull the dependency. I fixed `build.rs` so that we correctly refer to the system-installed abseil-cpp in case it's there.

I also rebased this branch of `main`. I had to squash it as I do that 🙏

skywhale commented:
@jacksongoode Could you pull the latest branch and see if the build succeeds?

The following commands should be sufficient for running the example `karaoke` program.

```
git clone https://github.com/tonarino/webrtc-audio-processing && cd webrtc-audio-processing
git submodule update --init --recursive
cargo run --example karaoke --features bundled --features derive_serde
```

jacksongoode commented:
> I removed it because the new build system using meson pulls the dependency automatically and installs it under the build directory, which we can reference.

This makes sense and also the build error as well. I cleaned and it works now! I'll close my PR.
